File: lib/cosmos/packets/telemetry.rb
# encoding: ascii-8bit

# Copyright 2014 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt

require 'cosmos/packets/packet_config'
require 'cosmos/ext/telemetry'

module Cosmos

  # Telemetry uses PacketConfig to parse the command and telemetry
  # configuration files. It contains all the knowledge of which telemetry packets
  # exist in the system and how to access them. This class is the API layer
  # which other classes use to access telemetry.
  #
  # This should not be confused with the Api module which implements the JSON
  # API that is used by tools when accessing the Server. The Api module always
  # provides Ruby primatives where the Telemetry class can return actual
  # Packet or PacketItem objects. While there are some overlapping methods between
  # the two, these are separate interfaces into the system.
  class Telemetry
    attr_accessor :config

    LATEST_PACKET_NAME = 'LATEST'.freeze

    # @param config [PacketConfig] Packet configuration to use to access the
    #   telemetry
    def initialize(config)
      @config = config
    end

    # (see PacketConfig#warnings)
    def warnings
      return @config.warnings
    end

    # @return [Array<String>] The telemetry target names (excluding UNKNOWN)
    def target_names
      result = @config.telemetry.keys.sort
      result.delete('UNKNOWN'.freeze)
      return result
    end

    # @param target_name [String] The target name
    # @return [Hash<packet_name=>Packet>] Hash of the telemetry packets for the given
    #   target name keyed by the packet name
    # def packets(target_name)

    # @param target_name [String] The target name
    # @param packet_name [String] The packet name. Must be a defined packet name
    #   and not 'LATEST'.
    # @return [Packet] The telemetry packet for the given target and packet name
    # def packet(target_name, packet_name)

    # @param target_name (see #packet)
    # @param packet_name (see #packet)
    # @return [Array<PacketItem>] The telemetry items for the given target and packet name
    def items(target_name, packet_name)
      return packet(target_name, packet_name).sorted_items
    end

    # @param target_name (see #packet)
    # @param packet_name (see #packet) The packet name.  LATEST is supported.
    # @return [Array<PacketItem>] The telemetry item names for the given target and packet name
    def item_names(target_name, packet_name)
      if LATEST_PACKET_NAME.casecmp(packet_name) == 0
        target_upcase = target_name.to_s.upcase
        target_latest_data = @config.latest_data[target_upcase]
        raise "Telemetry Target '#{target_upcase}' does not exist" unless target_latest_data
        item_names = target_latest_data.keys
      else
        tlm_packet = packet(target_name, packet_name)
        item_names = []
        tlm_packet.sorted_items.each {|item| item_names << item.name}
      end
      item_names
    end

    # @param target_name (see #packet)
    # @param packet_name [String] The packet name. 'LATEST' can also be given
    #   to specify the last received (or defined if no packets have been
    #   received) packet within the given target that contains the
    #   item_name.
    # @param item_name [String] The item name
    # @return [Packet, PacketItem] The packet and the packet item
    # def packet_and_item(target_name, packet_name, item_name)

    # Return a telemetry value from a packet.
    #
    # @param target_name (see #packet_and_item)
    # @param packet_name (see #packet_and_item)
    # @param item_name (see #packet_and_item)
    # @param value_type [Symbol] How to convert the item before returning.
    #   Must be one of {Packet::VALUE_TYPES}
    # @return The value. :FORMATTED and :WITH_UNITS values are always returned
    #   as Strings. :RAW values will match their data_type. :CONVERTED values
    #   can be any type.
    # def value(target_name, packet_name, item_name, value_type = :CONVERTED)

    # Set a telemetry value in a packet.
    #
    # @param target_name (see #packet_and_item)
    # @param packet_name (see #packet_and_item)
    # @param item_name (see #packet_and_item)
    # @param value The value to set in the packet
    # @param value_type (see #tlm)
    def set_value(target_name, packet_name, item_name, value, value_type = :CONVERTED)
      packet, _ = packet_and_item(target_name, packet_name, item_name)
      packet.write(item_name, value, value_type)
    end

    # @param target_name (see #packet_and_item)
    # @param item_name (see #packet_and_item)
    # @return [Array<Packet>] The latest (most recently arrived) packets with
    #   the specified target and item.
    def latest_packets(target_name, item_name)
      target_upcase = target_name.to_s.upcase
      item_upcase = item_name.to_s.upcase
      target_latest_data = @config.latest_data[target_upcase]
      raise "Telemetry target '#{target_upcase}' does not exist" unless target_latest_data
      packets = @config.latest_data[target_upcase][item_upcase]
      raise "Telemetry item '#{target_upcase} #{LATEST_PACKET_NAME} #{item_upcase}' does not exist" unless packets
      return packets
    end

    # @param target_name (see #packet_and_item)
    # @param item_name (see #packet_and_item)
    # @return [Packet] The packet with the most recent timestamp that contains
    #   the specified target and item.
    def newest_packet(target_name, item_name)
      # Handle LATEST_PACKET_NAME - Lookup packets for this target/item
      packets = latest_packets(target_name, item_name)

      # Find packet with newest timestamp
      newest_packet = nil
      newest_received_time = nil
      packets.each do |packet|
        received_time = packet.received_time
        if newest_received_time
          # See if the received time from this packet is newer.
          # Having the >= makes this method return the last defined packet
          # whether the timestamps are both nil or both equal.
          if received_time and received_time >= newest_received_time
            newest_packet = packet
            newest_received_time = newest_packet.received_time
          end
        else
          # No received time yet so take this packet
          newest_packet = packet
          newest_received_time = newest_packet.received_time
        end
      end
      return newest_packet
    end

    # Identifies an unknown buffer of data as a defined packet and sets the
    # packet's data to the given buffer. Identifying a packet uses the fields
    # marked as ID_ITEM to identify if the buffer passed represents the
    # packet defined. Incorrectly sized buffers are still processed but an
    # error is logged.
    #
    # Note: This affects all subsequent requests for the packet (for example
    # using packet) which is why the method is marked with a bang!
    #
    # @param packet_data [String] The binary packet data buffer
    # @param target_names [Array<String>] List of target names to limit the search. The
    #   default value of nil means to search all known targets.
    # @return [Packet] The identified packet with its data set to the given
    #   packet_data buffer. Returns nil if no packet could be identified.
    def identify!(packet_data, target_names = nil)
      identified_packet = nil

      target_names = target_names() unless target_names

      target_names.each do |target_name|
        target_packets = nil
        begin
          target_packets = packets(target_name)
        rescue RuntimeError
          # No telemetry for this target
          next
        end

        # Iterate through the packets and see if any represent the buffer
        target_packets.each do |packet_name, packet|
          if packet.identify?(packet_data)
            identified_packet = packet
            identified_packet.buffer = packet_data
            break
          end
        end
        break if identified_packet
      end
      return identified_packet
    end

    # Updates the specified packet with the given packet data. Raises an error
    # if the packet could not be found.
    #
    # Note: This affects all subsequent requests for the packet which is why
    # the method is marked with a bang!
    #
    # @param target_name (see #packet)
    # @param packet_name (see #packet)
    # @param packet_data (see #identify_tlm!)
    # @return [Packet] The packet with its data set to the given packet_data
    #   buffer.
    def update!(target_name, packet_name, packet_data)
      identified_packet = packet(target_name, packet_name)
      identified_packet.buffer = packet_data
      return identified_packet
    end

    # Assigns a limits change callback to all telemetry packets
    #
    # @param limits_change_callback
    def limits_change_callback=(limits_change_callback)
      @config.telemetry.each do |target_name, packets|
        packets.each do |packet_name, packet|
          packet.limits_change_callback = limits_change_callback
        end
      end
    end

    # Reads the specified list of items and returns their values and limits
    # state.
    #
    # @param item_array [Array<Array(String String String)>] An array
    #   consisting of [target name, packet name, item name]
    # @param value_types [Symbol|Array<Symbol>] How to convert the items before
    #   returning. A single symbol of {Packet::VALUE_TYPES}
    #   can be passed which will convert all items the same way. Or
    #   an array of symbols can be passed to control how each item is
    #   converted.
    # @return [Array, Array, Array] The first array contains the item values and the
    #   second their limits state, and the third their limits settings which includes
    #   the red, yellow, and green (if given) limits values.
    # def values_and_limits_states(item_array, value_types = :CONVERTED)

    # Iterates through all the telemetry packets and marks them stale if they
    # haven't been received for over the System.staleness_seconds value.
    def check_stale
      time = Time.now
      @config.telemetry.each do |target_name, target_packets|
        target_packets.each do |packet_name, packet|
          packet.set_stale if packet.received_time and (!packet.stale) and (time - packet.received_time > System.staleness_seconds)
        end
      end
    end

    # Clears the received_count value on every packet in every target
    def clear_counters
      @config.telemetry.each do |target_name, target_packets|
        target_packets.each do |packet_name, packet|
          packet.received_count = 0
        end
      end
    end

    # Resets metadata on every packet in every target
    def reset
      @config.telemetry.each do |target_name, target_packets|
        target_packets.each do |packet_name, packet|
          packet.reset
        end
      end
    end

    # Returns the first non-hidden packet
    def first_non_hidden
      @config.telemetry.each do |target_name, target_packets|
        next if target_name == 'UNKNOWN'
        target_packets.each do |packet_name, packet|
          return packet unless packet.hidden
        end
      end
      nil
    end

    # Returns an array with a "TARGET_NAME PACKET_NAME ITEM_NAME" string for every item in the system
    def all_item_strings(include_hidden = false, splash = nil)
      strings = []
      tnames = target_names()
      total = tnames.length.to_f
      tnames.each_with_index do |target_name, index|
        if splash
          splash.message = "Processing #{target_name} telemetry"
          splash.progress = index / total
        end

        ignored_items = System.targets[target_name].ignored_items

        packets(target_name).each do |packet_name, packet|
          # We don't audit against hidden or disabled packets
          next if !include_hidden and (packet.hidden || packet.disabled)
          packet.items.each_key do |item_name|
            # Skip ignored items
            next if !include_hidden and ignored_items.include? item_name
            strings << "#{target_name} #{packet_name} #{item_name}"
          end
        end
      end
      strings
    end

    # @return [Hash<String=>Packet>] Hash of all the telemetry packets
    #   keyed by the packet name.
    def all
      @config.telemetry
    end

  end # class Telemetry

end # module Cosmos

File: ext/cosmos/ext/telemetry/telemetry.c
/*
# Copyright 2014 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
*/

#include "ruby.h"
#include "stdio.h"

VALUE mCosmos;
VALUE cTelemetry;
VALUE cSystem;

static ID id_ivar_config = 0;
static ID id_method_telemetry = 0;
static ID id_method_read = 0;
static ID id_method_to_s = 0;
static ID id_method_intern = 0;
static ID id_method_get_item = 0;
static ID id_method_upcase = 0;
static ID id_method_newest_packet = 0;
static ID id_method_limits = 0;
static ID id_method_state = 0;
static ID id_method_limits_set = 0;
static ID id_method_values = 0;
static VALUE symbol_CONVERTED = Qnil;

/*
 * @param target_name [String] The target name
 *@return [Hash<packet_name=>Packet>] Hash of the telemetry packets for the given
 *  target name keyed by the packet name
 */
static VALUE packets(VALUE self, VALUE target_name) {
  VALUE target_packets = Qnil;
  VALUE upcase_target_name = Qnil;
  VALUE telemetry = Qnil;

  upcase_target_name = rb_funcall(target_name, id_method_to_s, 0);
  upcase_target_name = rb_funcall(upcase_target_name, id_method_upcase, 0);
  telemetry = rb_funcall(rb_ivar_get(self, id_ivar_config), id_method_telemetry, 0);
  target_packets = rb_hash_aref(telemetry, upcase_target_name);

  if (!(RTEST(target_packets))) {
    rb_raise(rb_eRuntimeError, "Telemetry target '%s' does not exist", RSTRING_PTR(upcase_target_name));
  }

  return target_packets;
}

/*
 * @param target_name [String] The target name
 * @param packet_name [String] The packet name. Must be a defined packet name
 *   and not 'LATEST'.
 *@return [Packet] The telemetry packet for the given target and packet name
 */
static VALUE packet(VALUE self, VALUE target_name, VALUE packet_name)
{
  VALUE packet = Qnil;
  VALUE target_packets = Qnil;
  VALUE upcase_target_name = Qnil;
  VALUE upcase_packet_name = Qnil;

  target_packets = packets(self, target_name);

  upcase_packet_name = rb_funcall(packet_name, id_method_to_s, 0);
  upcase_packet_name = rb_funcall(upcase_packet_name, id_method_upcase, 0);
  packet = rb_hash_aref(target_packets, upcase_packet_name);
  if (!(RTEST(packet))) {
    upcase_target_name = rb_funcall(target_name, id_method_to_s, 0);
    upcase_target_name = rb_funcall(upcase_target_name, id_method_upcase, 0);
    rb_raise(rb_eRuntimeError, "Telemetry packet '%s %s' does not exist", RSTRING_PTR(upcase_target_name), RSTRING_PTR(upcase_packet_name));
  }

  return packet;
}

/*
 * @param target_name (see #packet)
 * @param packet_name [String] The packet name. 'LATEST' can also be given
 *    to specify the last received (or defined if no packets have been
 *    received) packet within the given target that contains the
 *    item_name.
 * @param item_name [String] The item name
 * @return [Packet, PacketItem] The packet and the packet item
 */
static VALUE packet_and_item(VALUE self, VALUE target_name, VALUE packet_name, VALUE item_name)
{
  VALUE upcase_packet_name = Qnil;
  VALUE return_packet = Qnil;
  VALUE item = Qnil;
  VALUE return_value = Qnil;
  char * string_packet_name = NULL;

  upcase_packet_name = rb_funcall(packet_name, id_method_upcase, 0);
  string_packet_name = RSTRING_PTR(upcase_packet_name);
  if (strcmp(string_packet_name, "LATEST") == 0)
  {
    return_packet = rb_funcall(self, id_method_newest_packet, 2, target_name, item_name);
  }
  else
  {
    return_packet = packet(self, target_name, packet_name);
  }

  item = rb_funcall(return_packet, id_method_get_item, 1, item_name);

  return_value = rb_ary_new();
  rb_ary_push(return_value, return_packet);
  rb_ary_push(return_value, item);
  return return_value;
}

/*
 * Return a telemetry value from a packet.
 *
 * @param target_name (see #packet_and_item)
 * @param packet_name (see #packet_and_item)
 * @param item_name (see #packet_and_item)
 * @param value_type [Symbol] How to convert the item before returning.
 * Must be one of {Packet::VALUE_TYPES}
 * @return The value. :FORMATTED and :WITH_UNITS values are always returned
 * as Strings. :RAW values will match their data_type. :CONVERTED values
 * can be any type.
 */
static VALUE value(int argc, VALUE* argv, VALUE self)
{
  VALUE target_name = Qnil;
  VALUE packet_name = Qnil;
  VALUE item_name = Qnil;
  VALUE value_type = Qnil;
  VALUE result = Qnil;
  VALUE packet = Qnil;

  switch (argc)
  {
    case 3:
      target_name = argv[0];
      packet_name = argv[1];
      item_name = argv[2];
      value_type = symbol_CONVERTED;
      break;
    case 4:
      target_name = argv[0];
      packet_name = argv[1];
      item_name = argv[2];
      value_type = argv[3];
      break;
    default:
      /* Invalid number of arguments given */
      rb_raise(rb_eArgError, "wrong number of arguments (%d for 3..4)", argc);
      break;
  };

  result = packet_and_item(self, target_name, packet_name, item_name);
  packet = rb_ary_entry(result, 0);
  return rb_funcall(packet, id_method_read, 2, item_name, value_type);
}

/*
 * Reads the specified list of items and returns their values and limits
 * state.
 *
 * @param item_array [Array<Array(String String String)>] An array
 *   consisting of [target name, packet name, item name]
 * @param value_types [Symbol|Array<Symbol>] How to convert the items before
 *   returning. A single symbol of {Packet::VALUE_TYPES}
 *   can be passed which will convert all items the same way. Or
 *   an array of symbols can be passed to control how each item is
 *   converted.
 * @return [Array, Array, Array] The first array contains the item values, the
 *   second their limits state, and the third the limits settings which includes
 *   red, yellow, and green (if given) limits values.
 */
static VALUE values_and_limits_states(int argc, VALUE* argv, VALUE self) {
  VALUE item_array = Qnil;
  VALUE value_types = Qnil;
  VALUE items = Qnil;
  VALUE states = Qnil;
  VALUE settings = Qnil;
  VALUE entry = Qnil;
  VALUE target_name = Qnil;
  VALUE packet_name = Qnil;
  VALUE item_name = Qnil;
  VALUE value_type = Qnil;
  VALUE result = Qnil;
  VALUE return_value = Qnil;
  VALUE limits = Qnil;
  VALUE limits_set = Qnil;
  VALUE limits_values = Qnil;
  VALUE limits_settings = Qnil;
  long length = 0;
  long value_types_length = 0;
  int index = 0;

  switch (argc) {
    case 1:
      item_array = argv[0];
      value_types = symbol_CONVERTED;
      break;
    case 2:
      item_array = argv[0];
      value_types = argv[1];
      break;
    default:
      /* Invalid number of arguments given */
      rb_raise(rb_eArgError, "wrong number of arguments (%d for 1..2)", argc);
      break;
  };

  items = rb_ary_new();
  /* Verify items is a nested array */
  entry = rb_ary_entry(item_array, index);
  if (TYPE(entry) != T_ARRAY) {
    rb_raise(rb_eArgError, "item_array must be a nested array consisting of [[tgt,pkt,item],[tgt,pkt,item],...]");
  }
  states = rb_ary_new();
  settings = rb_ary_new();
  limits_set = rb_funcall(cSystem, id_method_limits_set, 0);
  length = RARRAY_LEN(item_array);
  if (TYPE(value_types) == T_ARRAY) {
    value_types_length = RARRAY_LEN(value_types);
    if (length != value_types_length) {
      rb_raise(rb_eArgError, "Passed %ld items but only %ld value types", length, value_types_length);
    }

    for (index = 0; index < length; index++) {
      entry = rb_ary_entry(item_array, index);
      target_name = rb_ary_entry(entry, 0);
      packet_name = rb_ary_entry(entry, 1);
      item_name = rb_ary_entry(entry, 2);
      value_type = rb_ary_entry(value_types, index);
      value_type = rb_funcall(value_type, id_method_intern, 0);

      result = packet_and_item(self, target_name, packet_name, item_name);
      rb_ary_push(items, rb_funcall(rb_ary_entry(result, 0), id_method_read, 2, item_name, value_type));
      limits = rb_funcall(rb_ary_entry(result, 1), id_method_limits, 0);
      rb_ary_push(states, rb_funcall(limits, id_method_state, 0));
      limits_values = rb_funcall(limits, id_method_values, 0);
      if (RTEST(limits_values)) {
        limits_settings = rb_hash_aref(limits_values, limits_set);
      } else {
        limits_settings = Qnil;
      }
      rb_ary_push(settings, limits_settings);
    }
  } else {
    value_type = rb_funcall(value_types, id_method_intern, 0);
    for (index = 0; index < length; index++) {
      entry = rb_ary_entry(item_array, index);
      target_name = rb_ary_entry(entry, 0);
      packet_name = rb_ary_entry(entry, 1);
      item_name = rb_ary_entry(entry, 2);

      result = packet_and_item(self, target_name, packet_name, item_name);
      rb_ary_push(items, rb_funcall(rb_ary_entry(result, 0), id_method_read, 2, item_name, value_type));
      limits = rb_funcall(rb_ary_entry(result, 1), id_method_limits, 0);
      rb_ary_push(states, rb_funcall(limits, id_method_state, 0));
      limits_values = rb_funcall(limits, id_method_values, 0);
      if (RTEST(limits_values)) {
        limits_settings = rb_hash_aref(limits_values, limits_set);
      } else {
        limits_settings = Qnil;
      }
      rb_ary_push(settings, limits_settings);
    }
  }

  return_value = rb_ary_new2(3);
  rb_ary_push(return_value, items);
  rb_ary_push(return_value, states);
  rb_ary_push(return_value, settings);
  return return_value;
}

/*
 * Initialize methods for Telemetry
 */
void Init_telemetry (void)
{
  id_ivar_config = rb_intern("@config");
  id_method_telemetry = rb_intern("telemetry");
  id_method_read = rb_intern("read");
  id_method_to_s = rb_intern("to_s");
  id_method_intern = rb_intern("intern");
  id_method_get_item = rb_intern("get_item");
  id_method_upcase = rb_intern("upcase");
  id_method_newest_packet = rb_intern("newest_packet");
  id_method_limits = rb_intern("limits");
  id_method_state = rb_intern("state");
  id_method_limits_set = rb_intern("limits_set");
  id_method_values = rb_intern("values");
  symbol_CONVERTED = ID2SYM(rb_intern("CONVERTED"));

  mCosmos = rb_define_module("Cosmos");
  cTelemetry = rb_define_class_under(mCosmos, "Telemetry", rb_cObject);
  rb_define_method(cTelemetry, "packets", packets, 1);
  rb_define_method(cTelemetry, "packet", packet, 2);
  rb_define_method(cTelemetry, "packet_and_item", packet_and_item, 3);
  rb_define_method(cTelemetry, "value", value, -1);
  rb_define_method(cTelemetry, "values_and_limits_states", values_and_limits_states, -1);

  cSystem = rb_define_class_under(mCosmos, "System", rb_cObject);
}
